src/aiida_quantumespresso/workflows/pw/base.py

```diff
@@ -482,6 +482,52 @@ def handle_vcrelax_converged_except_final_scf(self, calculation):
         self.results()  # Call the results method to attach the output nodes
         return ProcessHandlerReport(True, self.exit_codes.ERROR_IONIC_CONVERGENCE_REACHED_EXCEPT_IN_FINAL_SCF)
 
+    @process_handler(
+        priority=561,
+        exit_codes=[
+            PwCalculation.exit_codes.ERROR_IONIC_CYCLE_BFGS_HISTORY_FAILURE,
+            PwCalculation.exit_codes.ERROR_IONIC_CYCLE_BFGS_HISTORY_AND_FINAL_SCF_FAILURE,
+        ]
+    )
+    def handle_relax_recoverable_ionic_convergence_bfgs_history_error(self, calculation):
+        """Handle failure of the ionic minimization algorithm (BFGS).
+
+        When BFGS history fails, this can mean two things: the structure is close to the global minimum,
+        but the moves the algorithm wants to do are smaller than `trust_radius_min`, or the structure is
+        close to a local minimum (hard to detect). For the first, we restart with lowered trust_radius_min.
+        For the first case, one can lower the trust radius; for the second one, one can exploit a different
+        algorithm, e.g. `damp` (and `damp-w` for vc-relax).
+        """
+        trust_radius_min = self.ctx.inputs.parameters.setdefault('IONS', {}).setdefault('trust_radius_min', 1.0e-3)
+        calculation_type = self.ctx.inputs.parameters.setdefault('CONTROL', {}).setdefault('calculation', 'relax')
+        ion_dynamics = self.ctx.inputs.parameters.setdefault('IONS', {}).setdefault('ion_dynamics', 'bfgs')
+        if calculation_type == 'vc-relax':
+            cell_dynamics = self.ctx.inputs.parameters.setdefault('CELL', {}).setdefault('cell_dynamics', 'bfgs')
+
+        if calculation_type == 'relax' and ion_dynamics == 'bfgs':
+            self.ctx.inputs.parameters.setdefault('IONS', {})['ion_dynamics'] = 'damp'
+            action = 'bfgs history (ionic only) failure: restarting with `damp` dynamics.'
+        elif calculation_type == 'relax' and ion_dynamics == 'damp':
+            self.ctx.inputs.parameters.setdefault('IONS', {})['ion_dynamics'] = 'fire'
+            action = 'bfgs history (ionic only) failure: restarting with `fire` dynamics.'
+        elif trust_radius_min > 1.0e-4 and ion_dynamics == 'bfgs' and calculation_type == 'vc-relax':
```

Comment:
Nitpick: Just for consistency ^^

```suggestion
        calculation_type == 'vc-relax' and elif trust_radius_min > 1.0e-4 and ion_dynamics == 'bfgs':
```